File: src/libawkward/type/OptionType.cpp
// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/master/LICENSE

#include <string>
#include <sstream>

#include "awkward/type/UnknownType.h"
#include "awkward/type/ListType.h"
#include "awkward/type/RegularType.h"
#include "awkward/array/IndexedArray.h"
#include "awkward/Index.h"

#include "awkward/type/OptionType.h"

namespace awkward {
  OptionType::OptionType(const util::Parameters& parameters,
                         const std::string& typestr,
                         const TypePtr& type)
      : Type(parameters, typestr)
      , type_(type) { }

  std::string
  OptionType::tostring_part(const std::string& indent,
                            const std::string& pre,
                            const std::string& post) const {
    std::string typestr;
    if (get_typestr(typestr)) {
      return wrap_categorical(typestr);
    }

    std::stringstream out;
    if (parameters_empty()) {
      if (dynamic_cast<ListType*>(type_.get()) != nullptr  ||
          dynamic_cast<RegularType*>(type_.get()) != nullptr) {
        out << indent << pre << "option["
            << type_.get()->tostring_part(indent, "", "") << "]" << post;
      }
      else {
        out << indent << pre << "?"
            << type_.get()->tostring_part("", "", "") << post;
      }
    }
    else {
      out << indent << pre << "option["
          << type_.get()->tostring_part(indent, "", "") << ", "
          << string_parameters() << "]" << post;
    }
    return wrap_categorical(out.str());
  }

  const TypePtr
  OptionType::shallow_copy() const {
    return std::make_shared<OptionType>(parameters_, typestr_, type_);
  }

  bool
  OptionType::equal(const TypePtr& other, bool check_parameters) const {
    if (OptionType* t = dynamic_cast<OptionType*>(other.get())) {
      if (check_parameters  &&  !parameters_equal(other.get()->parameters())) {
        return false;
      }
      return type().get()->equal(t->type(), check_parameters);
    }
    else {
      return false;
    }
  }

  int64_t
  OptionType::numfields() const {
    return type_.get()->numfields();
  }

  int64_t
  OptionType::fieldindex(const std::string& key) const {
    return type_.get()->fieldindex(key);
  }

  const std::string
  OptionType::key(int64_t fieldindex) const {
    return type_.get()->key(fieldindex);
  }

  bool
  OptionType::haskey(const std::string& key) const {
    return type_.get()->haskey(key);
  }

  const std::vector<std::string>
  OptionType::keys() const {
    return type_.get()->keys();
  }

  const ContentPtr
  OptionType::empty() const {
    ContentPtr content = type_.get()->empty();
    Index64 index(0);
    return std::make_shared<IndexedOptionArray64>(Identities::none(),
                                                  parameters_,
                                                  index,
                                                  content);
  }

  const TypePtr
  OptionType::type() const {
    TypePtr out = type_;
    while (OptionType* t = dynamic_cast<OptionType*>(out.get())) {
      out = t->type_;
    }
    return out;
  }
}

File: src/awkward/_v2/types/optiontype.py
# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE

from __future__ import absolute_import

import awkward as ak
from awkward._v2.types.type import Type
from awkward._v2.types.regulartype import RegularType
from awkward._v2.types.listtype import ListType
from awkward._v2.types.uniontype import UnionType
from awkward._v2.forms.form import _parameters_equal


class OptionType(Type):
    def __init__(self, content, parameters=None, typestr=None):
        if not isinstance(content, Type):
            raise TypeError(
                "{0} 'content' must be a Type subclass, not {1}".format(
                    type(self).__name__, repr(content)
                )
            )
        if parameters is not None and not isinstance(parameters, dict):
            raise TypeError(
                "{0} 'parameters' must be of type dict or None, not {1}".format(
                    type(self).__name__, repr(parameters)
                )
            )
        if typestr is not None and not ak._util.isstr(typestr):
            raise TypeError(
                "{0} 'typestr' must be of type string or None, not {1}".format(
                    type(self).__name__, repr(typestr)
                )
            )
        self._content = content
        self._parameters = parameters
        self._typestr = typestr

    @property
    def content(self):
        return self._content

    def __str__(self):
        if self._typestr is not None:
            out = self._typestr

        else:
            params = self._str_parameters()
            if params is None:
                if not isinstance(self._content, (RegularType, ListType)):
                    out = "?{0}".format(str(self._content))
                else:
                    out = "option[{0}]".format(str(self._content))
            else:
                out = "option[{0}, {1}]".format(str(self._content), params)

        return self._str_categorical_begin() + out + self._str_categorical_end()

    def __repr__(self):
        args = [repr(self._content)] + self._repr_args()
        return "{0}({1})".format(type(self).__name__, ", ".join(args))

    def __eq__(self, other):
        if isinstance(other, OptionType):
            return (
                self._typestr == other._typestr
                and _parameters_equal(
                    self._parameters, other._parameters, only_array_record=True
                )
                and self._content == other._content
            )
        else:
            return False

    def simplify_option_union(self):
        if isinstance(self._content, UnionType):
            contents = []
            for content in self._content.contents:
                if isinstance(content, OptionType):
                    contents.append(
                        OptionType(
                            content.content,
                            ak._v2._util.merge_parameters(
                                self._parameters, content._parameters
                            ),
                            content._typestr
                            if self._typestr is None
                            else self._typestr,
                        )
                    )

                else:
                    contents.append(
                        OptionType(content, self._parameters, self._typestr)
                    )

            return UnionType(contents, self._content.parameters, self._content.typestr)

        else:
            return self
